2단계 - 지하철 노선 관리

hbkuk opened the pull request:
안녕하세요. 용주님!!

**2단계(지하철 노선 관리) 미션**에 대한 코드 리뷰 요청드립니다.
미션을 진행하면서 몇 가지 고민이 있었는데요. 이에 대해 용주님에 생각은 어떤지 궁금합니다.  
또, 궁금한 부분이 있었는데요. 설명 부탁드려도 괜찮으실까요?

## 고민
#### 테스트 검증 범위를 어디까지 두어야 할까요?
- 해당 미션에 대해서는 **모든 필드(`name`, `color`, `upStationId`, ...)에 대해서 검증**하는 방향으로 진행했습니다.
  - 각 테스트 메서드에서 `JsonPath`를 반환 받아서, `get` 혹은 `getList` 메서드를 통해 검증하는 것이 맞을까요?
    - 하지만, 가독성이 떨어지고, 검증 로직이 복잡해지는 느낌이 들었습니다.
#### **미션 진행 흐름(ATTD)은 아래와 같이 진행하는 것이 올바른 방향**일까요?
  1. 모든 인수 테스트 작성(리팩토링 진행 X)
  2. 인수 테스트가 성공할 때 까지 기능 구현(`Repository` -> `Service` -> `Controller`)
  3. 기능 리팩토링
  4. 인수 테스트 리팩토링
  5. 인수 테스트 최종 검토
#### **예를 들어, 전달받은 `stationLineId`에 해당하는 지하철 노선이 없을 경우  `Exception`을 Throw하는 것은 `Controller`가 아닌, `Service` 클래스에서 진행**하는 것이 괜찮은 방향일까요?
  - `Controller`는 요청이 제대로 오는지만 확인하고, Service로 부터 받은 데이터를 그대로 전달하는 역할만 하면 될까요?
#### 테스트 케이스의 수행 결과를 판별하는 메서드도 리팩토링 하는 것이 괜찮은 방향일까요?
- 만약, 분리한다면 어떻게 분리하는 것이 좋을지.. 힌트 주시면 감사드리겠습니다 ㅎㅎ;; 
<img width="820" alt="스크린샷 2024-01-28 212143" src="https://github.com/next-step/atdd-subway-map/assets/109803585/a6d8dde9-c9c6-4878-aa48-20c69161ef60">

## 질문
#### Service 클래스의 `updateStationLine` 메서드에 `@Transaction` 어노테이션이 없다면, 테스트 메서드에서 수정 전 데이터가 조회되며, 테스트가 실패합니다.
- `@Transaction` 어노테이션을 명시해줌으로써 해결은 했지만, 자세하게 확인해보지 않고 넘어갔습니다.
  - 이 부분은 제가 찾아보도록 하겠습니다.(혹시 아시는 내용이 있다면 설명 부탁드립니다. ㅎㅎ;;)

hbkuk committed:
docs: 1주차 리뷰 정리

hbkuk committed:
docs: 2주차 요구사항 정리

hbkuk committed:
test: 지하철 노선 생성 인수 테스트 작성

- feat: mock 데이터 추가

hbkuk committed:
test: 지하철 노선 목록 조회, 단건 조회 테스트 작성

hbkuk committed:
test: 지하철 노선 수정 테스트 작성

hbkuk committed:
test: 지하철 노선 삭제 테스트 작성

hbkuk committed:
feat: 지하철 노선 생성 기능 구현

hbkuk committed:
refactor: 요청 및 응답에서 도메인에 종속된 DTO 객체 사용

hbkuk committed:
fix: 기존 upStationId와 downStationId 필드의 데이터 타입을 Long에서 int으로 변경

- refactor: 지하철 노선 생성 테스트 메서드 리팩토링

hbkuk committed:
docs: ReadMe update

hbkuk committed:
refactor: 지하철 노선 생성 테스트 메서드 리팩토링

hbkuk committed:
refactor: 지하철 노선 목록 조회 테스트 메서드 리팩토링

- test: 테스트 메서드용 Mock 데이터 추가

hbkuk committed:
feat: 지하철 노선 조회 기능 구현

hbkuk committed:
fix: 기존 mock 데이터에 static 변수명을 Snake Case 형식으로 변경

hbkuk committed:
refactor: 지하철 노선 생성 테스트 메서드 리팩토링

hbkuk committed:
docs: ReadMe update

hbkuk committed:
feat: 지하철 노선 수정 기능 구현

hbkuk committed:
refactor: 지하철 노선 수정 기능 리팩토링

hbkuk committed:
feat: 지하철 노선 삭제 기능 구현

hbkuk committed:
docs: ReadMe update

hbkuk committed:
fix: 지하철 노선 삭제를 담당하는 서비스 메서드에 @Transactional 어노테이션 추가

hbkuk committed:
refactor: 지하철 노선 삭제 테스트 메서드 리팩토링

hbkuk committed:
docs: 완료된 TODO 주석 정리

hbkuk committed:
refactor: 기존 json key값에 해당하는 문자열을 상수로 분리

hbkuk committed:
docs: ReadMe update

hbkuk committed:
Merge branch 'hbkuk' into step3

testrace commented on README.md:
```diff
@@ -1,2 +1,52 @@
+        - Repository -> Service -> Controller
+     4. 기능 리팩토링
+     5. 인수 테스트 리팩토링 
+     6. 인수 테스트 최종 검토
```
작성하신 순서대로 진행하면 됩니다!
(넘버링이 조금 잘못된 것 같아요 😅  )

testrace commented on README.md:
```diff
@@ -1,2 +1,52 @@
+     5. 인수 테스트 리팩토링 
+     6. 인수 테스트 최종 검토
+  - throw Exception
+    - id에 해당하는 지하철 노선이 없을 경우, 예외를 Throw하는 위치
+      - 요청이 제대로 오는지 확인, 응답을 제대로 전달하는 것에만 집중하는 Controller에서 하는 것이 아닌 서비스에 역할일까?
```
이미 답을 알고 계시겠지만, 무엇이 맞다는 의견을 드리기 보다
Controller, Service 등의 계층을 분리하는 이유를 고민해 보시면 좋을 것 같습니다 😃

testrace commented on README.md:
```diff
@@ -1,2 +1,52 @@
+  - throw Exception
+    - id에 해당하는 지하철 노선이 없을 경우, 예외를 Throw하는 위치
+      - 요청이 제대로 오는지 확인, 응답을 제대로 전달하는 것에만 집중하는 Controller에서 하는 것이 아닌 서비스에 역할일까?
+  - `updateStationLine` 메서드는 @Transaction 어노테이션이 없다면, 조회 시 수정 전 데이터가 조회된다.(`deleteStationLine` 동일)
```
updateStationLine 메서드에 `@Transactional` 을 제거한 후 로그를 보시면 update 쿼리가 없는 것을 확인하실 수 있어요 
클래스에 적용된 `@Transactional` 애너테이션과 관련이 있는데요 
예시로 StationLineService의 모든 `@Transactional` 을 제거하고 테스트를 수행해 보셔도 좋을 것 같아요
JPA 사용 시 `@Transactional` 에 따라 내부 동작이 어떻게 되는지 찾아보시면 좋을 것 같아요

testrace commented on src/main/java/subway/dto/StationLineRequest.java:
```diff
@@ -0,0 +1,39 @@
+package subway.dto;
+
+public class StationLineRequest {
+    private final String name;
+    private final String color;
+    private final int upStationId;
+
+    private final int downStationId;
+
+    private final int distance;
```
```suggestion
    private final String name;
    private final String color;
    private final int upStationId;
    private final int downStationId;
    private final int distance;
```

hbkuk replied:
이 부분은 제가 놓쳤네요. ㅠ_ㅠ
감사합니다~!

testrace commented on src/main/java/subway/entity/StationLine.java:
```diff
@@ -0,0 +1,69 @@
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+
+@Entity
+public class StationLine {
+
+    public StationLine() {
+    }
```
클래스 구현 순서는 다음과 같이 중요한 순서로 위치하는게 좋을 것 같아요! 
참고 부탁드립니다 🙇🏻‍♀️

```
class  T {
    상수(static final) 또는 클래스 변수
    인스턴스 변수
    생성자
    팩토리 메소드
    메소드
    기본 메소드 (equals, hashCode, toString)
}
```

hbkuk replied:
이 부분도 제가 놓쳤네요.
언급해 주셔서 감사드립니다~!

testrace commented on src/main/java/subway/entity/StationLine.java:
```diff
@@ -0,0 +1,69 @@
+        return distance;
+    }
+
+    public StationLine update(String name, String color, int upStationId, int downStationId, int distance) {
+        this.name = name;
+        this.color = color;
+        this.upStationId = upStationId;
+        this.downStationId = downStationId;
+        this.distance = distance;
+
+        return this;
```
API 명세에는 name, color 두 필드만 수정하도록 되어 있는데요 
API 명세와 다르게 모든 필드를 수정할 수 있도록 구현하신 이유가 궁금합니다 🤔

hbkuk replied:
제가 API 명세를 제대로 확인하지 않았네요.(가장 기본인데 말이죠.. ㅜ_ㅜ)  
앞으로는 요구 조건을 확실하게 확인하고, 개발을 진행하도록 하겠습니다~!

testrace commented on src/main/java/subway/presentation/StationLineController.java:
```diff
@@ -0,0 +1,49 @@
+        return ResponseEntity.ok(stationLineService.findStationLineById(stationLineId));
+    }
+
+    @PatchMapping("/lines/{stationLineId}")
+    public ResponseEntity<StationLineResponse> updateStationLine(@PathVariable Long stationLineId,
+                                                                 @RequestBody StationLineRequest request) {
+        StationLineResponse stationLine = stationLineService.updateStationLine(stationLineId, request);
+        return ResponseEntity.created(URI.create("/lines/" + stationLine.getId())).body(stationLine);
+    }
```
이 부분에서도 http method, 응답 http status code가 API 명세와 다르네요 
미션 페이지에서 API 명세를 놓치신 것 같아요 😢

hbkuk replied:
😢

hbkuk replied:
수정하겠습니다.

testrace commented on src/test/java/subway/StationLineAcceptanceTest.java:
```diff
@@ -0,0 +1,281 @@
+        createStationLineRequest(mockRequest);
+        JsonPath allStationLine = findAllStationLines();
+
+        // then
+        assertAll(
+            () -> assertThat(allStationLine.getList(NAME_KEY)).containsAnyOf(mockRequest.getName()),
+            () -> assertThat(allStationLine.getList(COLOR_KEY)).containsAnyOf(mockRequest.getColor()),
+            () -> assertThat(allStationLine.getList(UP_STATION_ID_KEY)).containsAnyOf(mockRequest.getUpStationId()),
+            () -> assertThat(allStationLine.getList(DOWN_STATION_ID_KEY)).containsAnyOf(mockRequest.getDownStationId()),
+            () -> assertThat(allStationLine.getList(DISTANCE_KEY)).containsAnyOf(mockRequest.getDistance())
+        );
```
질문 주신 부분이네요 
인수 테스트는 사용자 스토리(시나리오)를 기반으로 인수 조건을 충족하는 테스트이기에 검증 범위가 정해져 있다고 보기엔 어려울 것 같아요 
모든 필드를 다 꺼내어서 비교하게 되면 지금처럼 오히려 가독성도 떨어지고 코드도 복잡해 지는데요
assertj의 usingRecursiveComparison 메서드를 활용하면 고민 하시는 부분에 도움이 될 수 있을 것 같아요

hbkuk replied:
앞으로 `usingRecursiveComparison` 메서드는 정말 유용하게 쓸 수 있을 것 같습니다. 👍
감사합니다.

testrace commented on src/test/java/subway/StationLineAcceptanceTest.java:
```diff
@@ -0,0 +1,281 @@
+     * When  생성한 지하철 노선을 삭제하면
+     * Then  해당 지하철 노선 정보는 삭제된다.
+     */
+    @DirtiesContext(methodMode = BEFORE_METHOD)
```
`@DirtiesContext`를 활용해 테스트를 격리해 주셨네요 👍 
테스트를 격리하기 위한 여러 방법 중  DirtiesContext를 적용하신 이유가 궁금합니다!
메서드가 아닌 클래스에도 적용 가능한 애너테이션인데요 클래스에 적용하면 어떨까요?

hbkuk replied:
제가 모르는게 너무 많네요.. 😢
용주님 덕분에 많이 배웠습니다!!👍

저는 스스로 찾아보는 습관이 필요한 것 같네요

testrace commented on src/test/java/config/fixtures/subway/StationLineMockData.java:
```diff
@@ -0,0 +1,38 @@
+
+public class StationLineMockData {
+
+    public static final String NAME_1 = "신분당선";
+    public static final String COLOR_1 = "bg-red-600";
+    public static final int UP_STATION_ID_1 = 1;
+    public static final int DOWN_STATION_ID_1 = 2;
+    public static final int DISTANCE_1 = 10;
```
fixture를 만들어 주셨네요 👍 
var1, var2 등 넘버링 변수명은 의미를 파악하기 어려워 권장하지 않으니 참고해 주세요
모든 필드를 변수로 만들기 보다 객체를 만들어도 좋을 것 같아요 
```java
public static final StationLineRequest 신분당선 = new StationLineRequest("신분당선", "bg-red-600", 1, 2, 10);
```

hbkuk replied:
넘버링 변수명은 만들면서 찝찝했는데.. 앞으로는 객체를 만들어서 사용하는 것이 좋겠네요 👍

testrace commented on src/test/java/subway/StationLineAcceptanceTest.java:
```diff
@@ -0,0 +1,281 @@
+        );
+    }
+
+    /**
+     * 지하철 노선 목록을 조회하고 jsonPath 반환
+     *
+     * @return 지하철 노선 목록을 나타내는 jsonPath 반환
+     */
+    private JsonPath findAllStationLines() {
+        return given().log().all()
+            .when()
+                .get("/lines")
+            .then().log().all()
+                .statusCode(HttpStatus.OK.value())
+                .extract().jsonPath();
+    }
```
RestAssured를 사용하는 메서드들을 모두 분리해 주셨네요 👍 
클래스로 분리해서 관리해도 좋을 것 같아요

hbkuk replied:
`StationLineRequest` 클래스를 만들어서, 정적 메서드로 분리해보았습니다.  😃
더 괜찮은 방법이 있을까요?

testrace replied:
개인적으로 StationLineRequest 는 DTO의 이름 같아요 
StationLineStpes, LineSteps 같은 이름을 선호하는 편입니다 😃

testrace reviewed (changes requested):
안녕하세요 병국님 😃 
커밋 로그와 readme에서 많은 고민을 하신게 느껴지네요 👍 
질문에 대한 답변과 코멘트 남겨두었는데요
확인해 주시고 다시 리뷰요청 부탁드려요

hbkuk committed:
fix: 미션 페이지의 API 명세 기반으로 지하철 노선 수정 관련 기능 수정

- docs: 로깅을 위한 JPA와 트랜잭션과 관련 설정 추가
- fix: 클래스 내부 변수, 메서드 선언 순서 수정 및 불필요한 공백 수정
- docs: 진행 순서 넘버링 수정

hbkuk committed:
refactor: 기존 Test Fixture 용도로 선언된 모호한 변수와 메서드를 명확한 객체로 변경

hbkuk committed:
test: 기존 테스트 검증 로직에서 usingRecursiveComparison 메서드 활용한 동등성 비교

- test: Test Fixture용 데이터 추가

hbkuk committed:
refactor: RestAssured를 사용하는 메서드를 StationLineRequest 클래스의 정적 메서드로 분리

hbkuk committed:
fix: 포맷팅 수정

testrace commented on src/main/java/subway/application/StationLineService.java:
```diff
@@ -53,10 +52,7 @@ public void deleteStationLine(Long stationLineId) {
-                request.getUpStationId(),
-                request.getDownStationId(),
-                request.getDistance()
+                request.getColor()
```
(optional)
validation 을 추가해 보셔도 좋을 것 같아요 😃

testrace commented on src/main/java/subway/entity/StationLine.java:
```diff
@@ -33,6 +22,23 @@ public StationLine(String name, String color, int upStationId, int downStationId
 
     private int distance;
 
+    public StationLine() {
+    }
```
jpa를 위한 생성자라면 proptected로 변경하면 어떨까요?

testrace commented on src/main/resources/application.properties:
```diff
@@ -1,2 +1,11 @@
+# Hibernate? ??? SQL ??? ??? ??
```
주석이 다 깨져서 나오네요 😢

testrace commented on src/test/java/subway/StationLineAcceptanceTest.java:
```diff
@@ -6,27 +6,21 @@
+import static subway.StationLineRequest.*;
 
 @DisplayName("지하철 노선 관련 기능")
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
```
클래스에 적용해 주셨군요 👍 
DirtiesContext 외에 다른 방법으로도 테스트를 격리해서 차이를 경험해 보시면 좋을 것 같아요

testrace commented on src/test/java/subway/StationLineRequest.java:
```diff
@@ -0,0 +1,88 @@
+     * @param request 지하철 요청 정보를 담은 객체
+     * @return REST Assured 기반으로 생성된 Response 객체
+     */
+    static ExtractableResponse<Response> createStationLineRequest(subway.dto.StationLineRequest request) {
```
클래스이름이 동일하니 패키지를 작성할 수 밖에 없으셨을 것 같아요
클래스명이 중복되지 않도록 변경하면 어떨까요?

testrace commented on src/test/java/subway/StationLineRequest.java:
```diff
@@ -0,0 +1,88 @@
+     * @param request       지하철 수정 정보를 담은 객체
+     * @param stationLineId 수정할 지하철 노선 ID
+     */
+    static void updateStationLineRequest(subway.dto.StationLineRequest request, Long stationLineId) {
```
fixture에 한글을 사용한 것처럼 메서드명에도 한글을 적용한다면 인수 테스트의 가독성에 도움이 될 것 같은데 어떻게 생각하시나요?
```suggestion
    static void 노선_수정_요청(subway.dto.StationLineRequest request, Long stationLineId) {
```

hbkuk replied:
아주 좋은 것 같습니다.  👍
다음 미션부터는 한글을 적극적으로 사용해보도록 하겠습니다.

testrace reviewed (approved):
안녕하세요 병국님 😃 
피드백 반영 잘해주셨네요 👍 
몇 가지 코멘트 남겨두었는데 확인해 주시고 다음 단계 진행해 주세요!
